[DDP] Uneven inputs: option to throw early

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56755 [DDP] Uneven inputs: option to throw early**

Rehash of https://github.com/pytorch/pytorch/pull/47488

Adds a flag to ddp join() context manager that enables throwing an error across all ranks when this flag is specified.

To do this, we implement the design in #47250. When running with this flag, we schedule an additional allreduce in the case that a joined rank needs to throw a StopIteration. In non-joined ranks forward pass, we match this allreduce and if at least one rank tells us to throw, we raise an error.

Tested by modifying existing tests, as well as adding additional tests validating that this works with SyncBatchNorm models and a model with custom collectives in the forward pass.

Due to the extra blocking allreduce introduced, there is an expected 2-3% perf gap when running with this flag. Benchmark results are available on internal diff. We will only recommend it for models with custom collective communication in the forward pass and consider performance improvements based on OSS user feedback.

Differential Revision: [D27958369](https://our.internmc.facebook.com/intern/diff/D27958369/)

rohan-varma committed:
[DDP] Uneven inputs: option to throw early

Rehash of https://github.com/pytorch/pytorch/pull/47488

Adds a flag to ddp join() context manager that enables throwing a
StopIteration across all ranks when this flag is specified.

To do this, we implement the design in #47250. When running with this flag, we schedule an additional allreduce in the case that a joined rank needs to throw a StopIteration. In non-joined ranks forward pass, we match this allreduce and if at least one rank tells us to throw, we raise a StopIteration.

Tested by modifying existing tests, as well as adding additional tests validating that this works with SyncBatchNorm models and a model with custom collectives in the forward pass.

Currently running perf benchmarks, will post when those are available, but we expect a small (~2%) perf reduction when enabling this feature due to the blocking allreduce. Hence we will only recommend it for models with collective comm.

Differential Revision: [D27958369](https://our.internmc.facebook.com/intern/diff/D27958369/)

[ghstack-poisoned]

rohan-varma commented:
Looking into test failures.

rohan-varma committed:
Update on "[DDP] Uneven inputs: option to throw early"


Rehash of https://github.com/pytorch/pytorch/pull/47488

Adds a flag to ddp join() context manager that enables throwing a
StopIteration across all ranks when this flag is specified.

To do this, we implement the design in #47250. When running with this flag, we schedule an additional allreduce in the case that a joined rank needs to throw a StopIteration. In non-joined ranks forward pass, we match this allreduce and if at least one rank tells us to throw, we raise a StopIteration.

Tested by modifying existing tests, as well as adding additional tests validating that this works with SyncBatchNorm models and a model with custom collectives in the forward pass.

Currently running perf benchmarks, will post when those are available, but we expect a small (~2%) perf reduction when enabling this feature due to the blocking allreduce. We expect this perf gap to be smaller for models with collective comm in the forward pass (which is the type of model this feature targets). Hence we will only recommend it for models with collective comm. in the forward pass and consider performance improvements based on OSS user feedback.

Differential Revision: [D27958369](https://our.internmc.facebook.com/intern/diff/D27958369/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Uneven inputs: option to throw early"


Rehash of https://github.com/pytorch/pytorch/pull/47488

Adds a flag to ddp join() context manager that enables throwing a
StopIteration across all ranks when this flag is specified.

To do this, we implement the design in #47250. When running with this flag, we schedule an additional allreduce in the case that a joined rank needs to throw a StopIteration. In non-joined ranks forward pass, we match this allreduce and if at least one rank tells us to throw, we raise a StopIteration.

Tested by modifying existing tests, as well as adding additional tests validating that this works with SyncBatchNorm models and a model with custom collectives in the forward pass.

Currently running perf benchmarks, will post when those are available, but we expect a small (~2%) perf reduction when enabling this feature due to the blocking allreduce. We expect this perf gap to be smaller for models with collective comm in the forward pass (which is the type of model this feature targets). Hence we will only recommend it for models with collective comm. in the forward pass and consider performance improvements based on OSS user feedback.

Differential Revision: [D27958369](https://our.internmc.facebook.com/intern/diff/D27958369/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Uneven inputs: option to throw early"


Rehash of https://github.com/pytorch/pytorch/pull/47488

Adds a flag to ddp join() context manager that enables throwing a
StopIteration across all ranks when this flag is specified.

To do this, we implement the design in #47250. When running with this flag, we schedule an additional allreduce in the case that a joined rank needs to throw a StopIteration. In non-joined ranks forward pass, we match this allreduce and if at least one rank tells us to throw, we raise a StopIteration.

Tested by modifying existing tests, as well as adding additional tests validating that this works with SyncBatchNorm models and a model with custom collectives in the forward pass.

Currently running perf benchmarks, will post when those are available, but we expect a small (~2%) perf reduction when enabling this feature due to the blocking allreduce. We expect this perf gap to be smaller for models with collective comm in the forward pass (which is the type of model this feature targets). Hence we will only recommend it for models with collective comm. in the forward pass and consider performance improvements based on OSS user feedback.

Differential Revision: [D27958369](https://our.internmc.facebook.com/intern/diff/D27958369/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Uneven inputs: option to throw early"


Rehash of https://github.com/pytorch/pytorch/pull/47488

Adds a flag to ddp join() context manager that enables throwing a
StopIteration across all ranks when this flag is specified.

To do this, we implement the design in #47250. When running with this flag, we schedule an additional allreduce in the case that a joined rank needs to throw a StopIteration. In non-joined ranks forward pass, we match this allreduce and if at least one rank tells us to throw, we raise a StopIteration.

Tested by modifying existing tests, as well as adding additional tests validating that this works with SyncBatchNorm models and a model with custom collectives in the forward pass.

Currently running perf benchmarks, will post when those are available, but we expect a small (~2%) perf reduction when enabling this feature due to the blocking allreduce. We expect this perf gap to be smaller for models with collective comm in the forward pass (which is the type of model this feature targets). Hence we will only recommend it for models with collective comm. in the forward pass and consider performance improvements based on OSS user feedback.

Differential Revision: [D27958369](https://our.internmc.facebook.com/intern/diff/D27958369/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Uneven inputs: option to throw early"


Rehash of https://github.com/pytorch/pytorch/pull/47488

Adds a flag to ddp join() context manager that enables throwing a
StopIteration across all ranks when this flag is specified.

To do this, we implement the design in #47250. When running with this flag, we schedule an additional allreduce in the case that a joined rank needs to throw a StopIteration. In non-joined ranks forward pass, we match this allreduce and if at least one rank tells us to throw, we raise a StopIteration.

Tested by modifying existing tests, as well as adding additional tests validating that this works with SyncBatchNorm models and a model with custom collectives in the forward pass.

Currently running perf benchmarks, will post when those are available, but we expect a small (~2%) perf reduction when enabling this feature due to the blocking allreduce. We expect this perf gap to be smaller for models with collective comm in the forward pass (which is the type of model this feature targets). Hence we will only recommend it for models with collective comm. in the forward pass and consider performance improvements based on OSS user feedback.

Differential Revision: [D27958369](https://our.internmc.facebook.com/intern/diff/D27958369/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Uneven inputs: option to throw early"


Rehash of https://github.com/pytorch/pytorch/pull/47488

Adds a flag to ddp join() context manager that enables throwing a
StopIteration across all ranks when this flag is specified.

To do this, we implement the design in #47250. When running with this flag, we schedule an additional allreduce in the case that a joined rank needs to throw a StopIteration. In non-joined ranks forward pass, we match this allreduce and if at least one rank tells us to throw, we raise a StopIteration.

Tested by modifying existing tests, as well as adding additional tests validating that this works with SyncBatchNorm models and a model with custom collectives in the forward pass.

Benchmark results are available on internal diff. We will only recommend it for models with collective comm. in the forward pass and consider performance improvements based on OSS user feedback.

Differential Revision: [D27958369](https://our.internmc.facebook.com/intern/diff/D27958369/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Uneven inputs: option to throw early"


Rehash of https://github.com/pytorch/pytorch/pull/47488

Adds a flag to ddp join() context manager that enables throwing a
StopIteration across all ranks when this flag is specified.

To do this, we implement the design in #47250. When running with this flag, we schedule an additional allreduce in the case that a joined rank needs to throw a StopIteration. In non-joined ranks forward pass, we match this allreduce and if at least one rank tells us to throw, we raise a StopIteration.

Tested by modifying existing tests, as well as adding additional tests validating that this works with SyncBatchNorm models and a model with custom collectives in the forward pass.

Benchmark results are available on internal diff. We will only recommend it for models with collective comm. in the forward pass and consider performance improvements based on OSS user feedback.

Differential Revision: [D27958369](https://our.internmc.facebook.com/intern/diff/D27958369/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Uneven inputs: option to throw early"


Rehash of https://github.com/pytorch/pytorch/pull/47488

Adds a flag to ddp join() context manager that enables throwing an error across all ranks when this flag is specified.

To do this, we implement the design in #47250. When running with this flag, we schedule an additional allreduce in the case that a joined rank needs to throw a StopIteration. In non-joined ranks forward pass, we match this allreduce and if at least one rank tells us to throw, we raise an error.

Tested by modifying existing tests, as well as adding additional tests validating that this works with SyncBatchNorm models and a model with custom collectives in the forward pass.

Benchmark results are available on internal diff. We will only recommend it for models with collective comm. in the forward pass and consider performance improvements based on OSS user feedback.

Differential Revision: [D27958369](https://our.internmc.facebook.com/intern/diff/D27958369/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Uneven inputs: option to throw early"


Rehash of https://github.com/pytorch/pytorch/pull/47488

Adds a flag to ddp join() context manager that enables throwing an error across all ranks when this flag is specified.

To do this, we implement the design in #47250. When running with this flag, we schedule an additional allreduce in the case that a joined rank needs to throw a StopIteration. In non-joined ranks forward pass, we match this allreduce and if at least one rank tells us to throw, we raise an error.

Tested by modifying existing tests, as well as adding additional tests validating that this works with SyncBatchNorm models and a model with custom collectives in the forward pass.

Benchmark results are available on internal diff. We will only recommend it for models with collective comm. in the forward pass and consider performance improvements based on OSS user feedback.

Differential Revision: [D27958369](https://our.internmc.facebook.com/intern/diff/D27958369/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Uneven inputs: option to throw early"


Rehash of https://github.com/pytorch/pytorch/pull/47488

Adds a flag to ddp join() context manager that enables throwing an error across all ranks when this flag is specified.

To do this, we implement the design in #47250. When running with this flag, we schedule an additional allreduce in the case that a joined rank needs to throw a StopIteration. In non-joined ranks forward pass, we match this allreduce and if at least one rank tells us to throw, we raise an error.

Tested by modifying existing tests, as well as adding additional tests validating that this works with SyncBatchNorm models and a model with custom collectives in the forward pass.

Benchmark results are available on internal diff. We will only recommend it for models with collective comm. in the forward pass and consider performance improvements based on OSS user feedback.

Differential Revision: [D27958369](https://our.internmc.facebook.com/intern/diff/D27958369/)

[ghstack-poisoned]

SciPioneer commented on torch/nn/parallel/distributed.py:
```diff
@@ -103,6 +103,7 @@ def _dump_DDP_relevant_env_vars():
 class _DDPUnevenInputsConfig(NamedTuple):
     ddp_join_enabled: bool
     ddp_join_divide_by_initial_world_size: bool
+    ddp_join_throw_on_early_termination: bool
```
Nit: May be worth noting the purpose of this field, such as when should a user flip it.

SciPioneer commented on torch/nn/parallel/distributed.py:
```diff
@@ -748,6 +751,14 @@ def forward(self, *inputs, **kwargs):
+                if self.ddp_uneven_inputs_config.ddp_join_throw_on_early_termination:
+                    zeros = torch.zeros(1, device=self.device)
+                    dist.all_reduce(zeros, group=self.process_group)
+                    should_throw_stop_iteration = zeros.item()
```
Are you expecting the `all_reduce` result to become 1 in any case? Shouldn't it be always 0s by adding up multiple 0s?

rohan-varma committed:
Update on "[DDP] Uneven inputs: option to throw early"


Rehash of https://github.com/pytorch/pytorch/pull/47488

Adds a flag to ddp join() context manager that enables throwing an error across all ranks when this flag is specified.

To do this, we implement the design in #47250. When running with this flag, we schedule an additional allreduce in the case that a joined rank needs to throw a StopIteration. In non-joined ranks forward pass, we match this allreduce and if at least one rank tells us to throw, we raise an error.

Tested by modifying existing tests, as well as adding additional tests validating that this works with SyncBatchNorm models and a model with custom collectives in the forward pass.

Due to the extra blocking allreduce introduced, there is an expected 2-3% perf gap when running with this flag. Benchmark results are available on internal diff. We will only recommend it for models with custom collective communication in the forward pass and consider performance improvements based on OSS user feedback.

Differential Revision: [D27958369](https://our.internmc.facebook.com/intern/diff/D27958369/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Uneven inputs: option to throw early"


Rehash of https://github.com/pytorch/pytorch/pull/47488

Adds a flag to ddp join() context manager that enables throwing an error across all ranks when this flag is specified.

To do this, we implement the design in #47250. When running with this flag, we schedule an additional allreduce in the case that a joined rank needs to throw a StopIteration. In non-joined ranks forward pass, we match this allreduce and if at least one rank tells us to throw, we raise an error.

Tested by modifying existing tests, as well as adding additional tests validating that this works with SyncBatchNorm models and a model with custom collectives in the forward pass.

Due to the extra blocking allreduce introduced, there is an expected 2-3% perf gap when running with this flag. Benchmark results are available on internal diff. We will only recommend it for models with custom collective communication in the forward pass and consider performance improvements based on OSS user feedback.

Differential Revision: [D27958369](https://our.internmc.facebook.com/intern/diff/D27958369/)

[ghstack-poisoned]
